api.md

```diff
@@ -0,0 +1,215 @@
+
+### API Endpoints
+
+An "endpoint" is a combination of two things:
+
+* The verb (e.g. `GET` or `POST`)
+* The URL path (e.g. `/articles`)
+
+Information can be passed to an endpoint in either of two ways:
+
+* The URL query string (e.g. `?year=2014`)
+* HTTP headers (e.g. `X-Api-Key: my-key`)
+
+When people say "RESTful" nowadays, they really mean designing simple, intuitive endpoints that represent unique functions in the API.
+
+Generally speaking:
+
+* **Avoid single-endpoint APIs.** Don't jam multiple operations into the same endpoint with the same HTTP verb.
+* **Prioritize simplicity.** It should be easy to guess what an endpoint does by looking at the URL and HTTP verb, without needing to see a query string.
+* Endpoint URLs should advertise resources, and **avoid verbs**.
+
+Some examples of these principles in action:
+
+* [FBOpen API documentation](https://18f.github.io/fbopen/)
+* [OpenFDA example query](https://open.fda.gov/api/reference/#example-query)
+* [Sunlight Congress API methods](https://sunlightlabs.github.io/congress/#using-the-api)
+
+### Just use JSON
+
+[JSON](https://en.wikipedia.org/wiki/JSON) is an excellent, widely supported transport format, suitable for many web APIs.
+
+Supporting JSON and only JSON is a practical default for APIs, and generally reduces complexity for both the API provider and consumer.
+
+General JSON guidelines:
+
+* Responses should be **a JSON object** (not an array). Using an array to return results limits the ability to include metadata about results, and limits the API's ability to add additional top-level keys in the future.
+* **Don't use unpredictable keys**. Parsing a JSON response where keys are unpredictable (e.g. derived from data) is difficult, and adds friction for clients.
+* **Use consistent case for keys**. Whether you use `under_score` or `CamelCase` for your API keys, make sure you are consistent.
+
+### Use a consistent date format
+
+And specifically, [use ISO 8601](https://xkcd.com/1179/), in UTC.
+
+For just dates, that looks like `2013-02-27`. For full times, that's of the form `2013-02-27T10:00:00Z`.
+
+This date format is used all over the web, and puts each field in consistent order -- from least granular to most granular.
+
+
+### API Keys
+
+These standards do not take a position on whether or not to use API keys.
+
+But _if_ keys are used to manage and authenticate API access, the API should allow some sort of unauthenticated access, without keys.
+
+This allows newcomers to use and experiment with the API in demo environments and with simple `curl`/`wget`/etc. requests.
+
+Consider whether one of your product goals is to allow a certain level of normal production use of the API without enforcing advanced registration by clients.
+
+
+### Error handling
+
+Handle all errors (including otherwise uncaught exceptions) and return a data structure in the same format as the rest of the API.
```

Comment:
I don't think our guidelines need to take a position on how rich the errors should be. The two things I think we should emphasize are:

1. The error should use the appropriate HTTP status code (which I would prioritize *above* returning data in JSON).
2. The error should return some data in same format as a non-error, i.e. JSON if we're following the recommendation above. It can be as rich or not as necessary.

I would probably like this section reorganized in that order or prioritization — with using appropriate  HTTP status codes first. Currently that's only tangentially hinted at in the paragraph following the example.